### ISSUE TITLE
SPI frequency failing above 10mhz frequency

### ISSUE THREAD

**Issue by Makuna** [esp8266/Arduino#5683]
### Basic Infos

- [ X] This issue complies with the [issue POLICY doc](https://github.com/esp8266/Arduino/blob/master/POLICY.md).
- [X ] I have read the documentation at [readthedocs](https://arduino-esp8266.readthedocs.io/en/latest) and the issue is not addressed there.
- [X ] I have tested that the issue is present in current master branch (aka latest git).
- [X ] I have searched the issue tracker for a similar issue.
- [ ] If there is a stack dump, I have decoded it.
- [X ] I have filled out all fields below.

#### Platform

- Hardware:			ESP-12
- Core Version:			latest
- Development Env:		Arduino IDE
- Operating System:		Windows

### Settings in IDE

- Module:  			LOLIN (WEMOS) D! R@ & Mini
- Flash Mode:			option not shown
- Flash Size: 		4MB no spiffs
- lwip Variant:   v2 Lower Memory
- Reset Method:		option not shown
- Flash Frequency:	option not shown
- CPU Frequency:		80Mhz
- Upload Using:		SERIAL
- Upload Speed:   921600

### Problem Description
Setting the SPI fequency above 10mhz (11mhz will acutally use 10mhz as part of the code that finds best match) will fail to send the data correctly.

I would either expect the max supported frequency to limit it 10mhz or if it truely can support higher then it should be fixed to do so.

This can be tested using the below sketch, then in the Serial monitor, type "+++++++++++++" to set the frequency to 14mhz; then a "w" to writeBytes 7 bytes of data out on SPI.  Capture the SPI data using a logic probe.

You will see something akin to this
![image](https://user-images.githubusercontent.com/7972357/51808254-76efe880-2246-11e9-9d30-06b2b140f606.png)

What should be seen is something like this (set to 10mhz)
![image](https://user-images.githubusercontent.com/7972357/51808277-af8fc200-2246-11e9-8517-ac8b86f42d1c.png)

### [MCVE](https://stackoverflow.com/help/mcve) Sketch

```cpp
#include <SPI.h>

const uint8_t PIN_SS = 2;
uint8_t c_sendBuffer[] = {0xf0, 0x55, 0x55, 0x55, 0x55, 0x55, 0x0f};
uint32_t g_frequency = 1000000L;
uint8_t* g_sendBuffer;
uint8_t* g_receiveBuffer;

#define countof(a)  (sizeof(a)/sizeof(a[0]))

void InitSPI() {
  SPI.begin();

  pinMode(PIN_SS, OUTPUT);
  digitalWrite(PIN_SS, HIGH);
}

void setup() {
  // put your setup code here, to run once:
    Serial.begin(115200);
    while (!Serial); // wait for serial attach

    Serial.println();
    Serial.println("Initializing...");
    Serial.flush();

    InitSPI();

    g_sendBuffer = new uint8_t[sizeof(c_sendBuffer)];
    memcpy(g_sendBuffer, c_sendBuffer, sizeof(c_sendBuffer));

    g_receiveBuffer = new uint8_t[sizeof(c_sendBuffer)];
    memset(g_receiveBuffer, 0, sizeof(c_sendBuffer));
    
    Serial.print("0x");
    Serial.println((uint32_t)(&c_sendBuffer[0]), HEX);

    Serial.print("0x");
    Serial.println((uint32_t)(g_sendBuffer), HEX);
    
    Serial.println();
    Serial.println("Running...");

}

class SpiTransaction
{
  public:
      SpiTransaction(uint32_t frequency) {
        SPI.beginTransaction(SPISettings(frequency, MSBFIRST, SPI_MODE0));
        digitalWrite(PIN_SS, LOW);
      }
      ~SpiTransaction() {
        digitalWrite(PIN_SS, HIGH);
        SPI.endTransaction();
      }
};

void printNewFrequency() {
        Serial.println();
        Serial.print("requency set to ");
        Serial.println(g_frequency);
}

void loop() {
  if (Serial.available()) {
    int cmd = Serial.read();
    if (cmd != -1) {
      switch ((char)cmd) {
        case '+':
        g_frequency += 1000000L;
        printNewFrequency();
        break;
        
        case '-':
        g_frequency -= 1000000L;
        printNewFrequency();
        break;
        
        /* 2.4.0 didn't have transfer */
        case 't':
        {
          SpiTransaction spiTransaction(g_frequency);
          SPI.transfer(g_sendBuffer, sizeof(c_sendBuffer));
        }
        // transfer overwrites buffer with received data, even if
        // most cases this is just 0xff
        memcpy(g_sendBuffer, c_sendBuffer, sizeof(c_sendBuffer));
        Serial.println("used transfer");
        break;
        /* */
        
 #if defined(ARDUINO_ARCH_ESP8266) || defined(ARDUINO_ARCH_ESP32)         
        case 'w':
        {
          SpiTransaction spiTransaction(g_frequency);
          SPI.writeBytes(g_sendBuffer, sizeof(c_sendBuffer));
        }   
        Serial.println("used writeBytes");
        break;
       
        case 'b':
        {
          SpiTransaction spiTransaction(g_frequency);
          SPI.transferBytes(g_sendBuffer, g_receiveBuffer, sizeof(c_sendBuffer));
        } 
        Serial.println("used transferBytes");
        break;

        case 'B':
        {
          SpiTransaction spiTransaction(g_frequency);
          SPI.transferBytes(g_sendBuffer, NULL, sizeof(c_sendBuffer));
        } 
        Serial.println("used transferBytes");

        break;
#endif        
        
        default:
        break;
      }
    }
  }
}
```

### Debug Messages

Internally, SPI best match code will pick a final freq = 14000000; and the real freq = 13333333

**Comment by Makuna** [comment]
If I could get a confirmation from another person with a Logic Analyzer on the results it would be great.  There is a little doubt that mine may not work well near this range.

**Comment by earlephilhower** [comment]
Unfortunately I've only access to one of those knock off USB analyzers w/24MHz sampling rate, so can't go above 12MHz.  However, in my SDFS work (using SPI to talk to the SD card) ~10Mhz was the highest I was able to get it to work.

The SPI transfer itself is handled in custom HW on chip, so if you look at libraries/SPI you'll see for transfers all we're doing is loading a FIFO that the HW pumps out.

The clock generation HW is set up in SPIClass::setFrequency.

**Comment by devyte** [comment]
@Makuna [some reading from Espressif](https://www.espressif.com/sites/default/files/documentation/esp8266-technical_reference_en.pdf), if you haven't already.
Apparently, the speeds differ quite a bit depending on the SPI mode used.

**Comment by devyte** [comment]
Also, it might be worth investigating how the [lua nodemcu spi](https://nodemcu.readthedocs.io/en/master/en/modules/spi/#spisetup) is handled. Notice the clock divider.

**Comment by devyte** [comment]
Related to my previous comment, [this](https://github.com/nodemcu/nodemcu-firmware/pull/1283) might be worth a look for contrasting their register setup to ours.

**Comment by Makuna** [comment]
I just ran through some tests with a logic analyser that can do 50mhz SPI.
A few of note 
freq = 20000000  real freq = 20000000
freq = 24000000  real freq = 20000000
freq = 30000000  real freq = 26666666
freq = 50000000  real freq = 40000000
all frequency requests >50mhz and <80mhz would always pick a real freq of 40mhz; a clk division of two.
At 80mhz requests and above, the code drops through a different path (so no debg output and the data showed as garbage).  This could be that it was really garbage, or that the analyzer could not handle 80mhz as it is out of its spec.  I suspect it was sending at 80mhz.
I will do a little investigation on the different code path; and if it is setting to 80mhz (of which it should be able to do in master mode) then this issue will get closed as it seems to be confusion caused by my analyzer not being able to handle the speeds it supports.

**Comment by Bodmer** [comment]
I use the ESP8266 to drive TFT displays and it works fine at 80MHz even though this is way beyond the spec for the TFT.

I suspect this is an issue with the test equipment being used, this is evident from the uneven SCK mark space ratio, essentially the sampling clock is aliasing with the SPI SCK signal so that is not a true picture of the waveform. Bear in mind the decoding with test equipment is based on a sampling rather than edge clocking basis and therefore the sample rate must be much higher than the SPI rate, so it is the test equipment is failing to decode the stream.

**Comment by Makuna** [comment]
review the code for >= 80mhz SPI frequency.  It shortcuts a bunch of the existing code for finding best fit as it is the max freq supported, thus the missing output I noticed.  I confirmed it is setting the clock to the max of 80mhz.
An interesting thing to note, the symbol `ESP8266_CLOCK` is the peripheral clock and not the CPU clock, so it is always 80mhz not matter what you set the board option to.  The name could be more descriptive to this; but it is accurate.
